docs(examples): simplify the barchart example

joshka opened the pull request:
The `barchart` example has been split into two examples: `barchart` and
`barchart-grouped`. The `barchart` example now shows a simple barchart
with random data, while the `barchart-grouped` example shows a grouped
barchart with fake revenue data.

This simplifies the examples a bit so they don't cover too much at once.

- Simplify the rendering functions
- Fix several clippy lints that were marked as allowed

EdJoPaTo commented on examples/barchart.rs:
```diff
@@ -19,59 +19,51 @@ use std::{
-    terminal::{disable_raw_mode, enable_raw_mode, EnterAlternateScreen, LeaveAlternateScreen},
-};
+use crossterm::event::{self, Event, KeyCode};
+use rand::prelude::*;
```
I dislike prelude as it hides what exactly is used and where it comes from.

Using `ratatui::prelude::*` in ratatui examples is not ideal in my opinion as it hides what exactly is used, but it seems like a wanted feature so ok. Using prelude from other crates only increases the confusion what's from where and what is used here.

As I seem to have different ideas about good Rust code I stopped looking/caring about this PR after seeing this. As this PR came up again I post this as a head up.

joshka replied:
`rand::prelude` is used in their docs (first example in https://docs.rs/rand/latest/rand/)

My take on the use of a prelude in Ratatui is that it removes a bunch of unnecessary boilerplate code and seems to have little to no downside. Can you maybe help me understand what problems you see with using it? The first one you mentioned that there's some confusion about what types are from which libraries, but that's something that any modern IDE or editor with LSP supports. Is there a more concrete example of this that you could share where this has caused problems?

What are the other parts of this that you see as not good rust code?

EdJoPaTo replied:
Can we be more explicit here what is actually used of rand? I have used rand in different repos but never used their prelude and didn't notice they have one. Looking at my own code, probably this, but I haven't tested it:

```suggestion
use rand::Rng;
```

joshka replied:
Rand docs use the prelude - that's their blessed path, so I'd stick with it for this instance.

joshka commented:
I made a bunch of changes:
- removed the prelude
- revamped the way the data for the companies was stored to simplify the conversion into bars (i.e. it's stored as months containing revenues for each company rather than companies containing revenues for each month
- moved a bunch of things around
- tightened up pretty much everything
- added comments where they would normally be needed.

joshka commented:
The idea behind this is to do similar changes to all the examples, and to make them take this form.

joshka commented on examples/barchart.rs:
```diff
@@ -13,286 +13,267 @@
-                    return Ok(());
+    /// Run the application
+    fn run(mut self, terminal: &mut Terminal<impl Backend>) -> color_eyre::Result<()> {
+        let mut last_tick = Instant::now();
+        loop {
+            terminal.draw(|frame| frame.render_widget(&self, frame.size()))?;
+
+            let timeout = TICK_RATE.saturating_sub(last_tick.elapsed());
+            if event::poll(timeout)? {
+                if let Event::Key(key) = event::read()? {
+                    if key.code == KeyCode::Char('q') {
+                        break;
+                    }
                 }
             }
+            if last_tick.elapsed() >= TICK_RATE {
+                // only update the data every 250ms
+                self.update_data();
+                last_tick = Instant::now();
+            }
         }
```
I'm not sure that I understand the question. The tick is a mechanism for ensuring that the data updates every 250ms. The method that updates the data is directly below this called update_data.

joshka replied:
Perhaps naming it something other than tick would help e.g. "last_update" UPDATE_RATE or something similar?

joshka replied:
Changed this a bit in https://github.com/ratatui-org/ratatui/pull/1079/commits/657ab84a636a2019f603632fe347434e7df01835

joshka commented on examples/barchart.rs:
```diff
@@ -13,286 +13,267 @@
+        self.vertical_revenue_barchart().render(left, buf);
+        self.horizontal_revenue_barchart().render(right, buf);
+        self.legend().render(left_legend_area, buf);
+        self.legend().render(right_legend_area, buf);
```
`buf` is the convention that we use throughout ratatui. I think changing that is a ship that sailed long ago, and not one to change in an example like this. Consistency is more important here than using a more full name I think. (I say this as someone that dislikes abbreviations generally)

If we did change this, I'd want to change it everywhere all at once, in a major release so that all our docs / website code / examples all use buffer.

EdJoPaTo replied:
> I think changing that is a ship that sailed long ago

Why?

The name in the trait does not need to be the name in the implementation. Only the types need to match. Nothing is breaking by changing that.

> If we did change this, I'd want to change it everywhere all at once, in a major release so that all our docs / website code / examples all use buffer.

This requires way more effort than just doing it when visiting it at some place, changing it while there. Seems to be a high bar for something that is relatively simple and painless.

joshka replied:
> Why?

Because it's used as buf conventionally everywhere, in every widget and every example (221 usages in ratatui), in code that's built with ratatui (GitHub search - 1.5K files) There are 14 files that use buffer instead of buf as a function parameter name.

> Nothing is breaking by changing that.

This is not about whether or not it would be breaking, this is about not making a change that isn't necessary. Buf as an abbreviation is well accepted. I didn't like it to begin with either, but my heuristic for choosing what to write weights convention over opinion in this.

I could get behind changing this across the board to drive things towards a new convention more broadly, but not piecemeal like this just in one example (or even all of them). I do agree with you that buffer is the right name for this parameter generally.

joshka commented on examples/barchart.rs:
```diff
@@ -13,286 +13,267 @@
+impl App<'_> {
+    /// Create a bar chart with the data from the `data` field.
+    fn main_barchart(&self) -> BarChart<'_> {
+        BarChart::default()
+            .block(Block::bordered().title("Vertical Grouped"))
+            .data(BarGroup::default().bars(&self.data))
```
yup - agreed, but it should actually be:
```rust
BarChart::new(vec![BarGroup::new(&self.data)])...
```
Or something like that.

Either way, this is a deeper conversation that I'd prefer to do orthogonality to the changes for examples (I picked BarChart based on alphabetic order not for any other reason).

joshka commented on examples/barchart.rs:
```diff
@@ -13,286 +13,267 @@
+            .block(Block::bordered().title("Vertical Grouped"))
+            .data(BarGroup::default().bars(&self.data))
+            .bar_width(5)
+            .bar_style(Style::default().fg(Color::Yellow))
+            .value_style(Style::default().fg(Color::Black).bg(Color::Yellow))
```
Same - orthogonal comment as above - BarGroup doesn't have a const constructor yet.
Also the style methods are non-const due to the `Into<Style>` calls (which won't be removed).

Applying this change however.

joshka replied:
Done in 959b2a06e

joshka commented on examples/common.rs:
```diff
@@ -0,0 +1,85 @@
+    config::{EyreHook, HookBuilder, PanicHook},
+    eyre::{self},
+};
+use crossterm::{
+    execute,
+    terminal::{
+        disable_raw_mode, enable_raw_mode, Clear, ClearType, EnterAlternateScreen,
+        LeaveAlternateScreen,
+    },
+};
+use ratatui::{backend::CrosstermBackend, Terminal};
+
+/// Initialize the terminal by enabling raw mode and entering the alternate screen.
+pub fn init_terminal() -> io::Result<Terminal<CrosstermBackend<Stdout>>> {
+    enable_raw_mode()?;
+    execute!(stdout(), EnterAlternateScreen)?;
+    let backend = CrosstermBackend::new(stdout());
+    let terminal = Terminal::new(backend)?;
+    Ok(terminal)
+}
+
+/// Restore the terminal by leaving the alternate screen and disabling raw mode.
+pub fn restore_terminal() -> color_eyre::Result<()> {
+    disable_raw_mode()?;
+    execute!(
+        stdout(),
+        LeaveAlternateScreen,
+        Clear(ClearType::FromCursorDown),
+    )?;
+    Ok(())
+}
```
Yes, but should be discussed orthogonality to this. See #280 for background.

joshka commented on examples/common.rs:
```diff
@@ -0,0 +1,85 @@
+    Ok(())
+}
+
+/// This replaces the standard `color_eyre` panic and error hooks with hooks that restore the
+/// terminal before printing the panic or error.
+pub fn install_hooks() -> color_eyre::Result<()> {
+    let (panic_hook, eyre_hook) = HookBuilder::default().into_hooks();
+    install_panic_hook(panic_hook);
+    install_error_hook(eyre_hook)?;
+    Ok(())
+}
+
+/// Install a panic hook that restores the terminal before printing the panic.
+fn install_panic_hook(panic_hook: PanicHook) {
+    // convert from a color_eyre PanicHook to a standard panic hook
+    let panic_hook = panic_hook.into_panic_hook();
+    panic::set_hook(Box::new(move |panic_info| {
+        let _ = restore_terminal();
+        panic_hook(panic_info);
+    }));
+}
+
+/// Install an error hook that restores the terminal before printing the error.
+fn install_error_hook(eyre_hook: EyreHook) -> color_eyre::Result<()> {
+    // convert from a color_eyre EyreHook to a standard eyre hook
+    let eyre_hook = eyre_hook.into_eyre_hook();
+    eyre::set_hook(Box::new(move |error| {
+        let _ = restore_terminal();
+        eyre_hook(error)
+    }))?;
+    Ok(())
+}
```
We can, but I'm going to go with this being extremely useful based on seeing many different apps that crash for one reason or another (or return errors). When we don't use color_eyre or anyhow, then we have to specifically do:

```rust
let result = app.run();
restore_terminal()?;
result
```
Which is janky.

We're building a standard and opinionated approach here that many apps will build on top of. Ratatui is used by new rust users when they're learning rust, and it's useful to just do the right thing. Color eyre is a reasonable opinion that we can build on, is documented well in our website as a pattern to follow in the counter tutorial and the how-to pages.

This sets new users up to succeed automatically (when their app fails, they have a good way of understanding why upfront).

EdJoPaTo replied:
This can work perfectly without an additional crate:

https://github.com/ratatui-org/ratatui/blob/9bd89c218afb1f3999dce1bfe6edea5b7442966d/examples/panic.rs#L52-L57

joshka replied:
As an aside, I think we probably should make some sort of easy way to integrate color_eyre into an app via crate feature flag instead of manually including it. E.g. perhaps something like:

```rust
ratatui::backend::crossterm::install_hooks()?;
let terminal = ratatui::backend::crossterm::terminal_on_stdout()?; // or something similar
```

(Note: the mechanism for enabling raw mode / entering the alternate screen is quite a bit different for each of the backends in ways that are generally incompatible with each other.)

I think you might be right that it would be nice to re-export crossterm instead of having to manually add that to your deps. Definitely a thing to consider outside of this PR though.

A concrete example of inconsistent approaches leading to user confusion can be found in the live streams that this newer rust user is running going through their learnings on Ratatui. https://www.youtube.com/@GoRustLang (see the recent livestreams - they're worth watching at 2x+ speed. In them the author actually hit questions about error handling and saw problems with it when they crash their app. They originally also learnt from the JSON tutorial, which takes a frame-centric approach to building apps rather than a widget approach. This made for confusion about how and where to get a buffer from when they were reading a widget-centric example. What this really showed me is that people learn non-linearly, and making sure there's consistency across the various examples and docs makes it easy for users to jump around and not be looking at drastic differences. It's something that I'd hypothesized that it would be a thing, but the series really showed that by example.

kdheepak replied:
I am in agreement about the non-linear learning and support consistency.

I'm personally a fan of having nicer panic hooks and have grown fond of the output from color-eyre. But also I see the value in showing only what is related to `ratatui`.

To that end, would having a `minimalist-example.rs` be useful?

joshka replied:
> This can work perfectly without an additional crate:

That's definitely far from "perfect":

- doesn't handle errors, only panics. A result return type is easy to overlook and forget to handle in the main method (e.g. just naively using the `?` operator)
- doesn't provide error messages that are anywhere close to those provided out of the box by color_eyre
- doesn't provide an abstraction that makes it easy to avoid talking about boxing errors
- error messages have the failing line of code somewhere in the stack trace, but require searching to work out the cause instead of being the first element

The benefits of color-eyre are all the sorts of things that help newer users fall into the pit of success rather than being tripped up.

Compare:
standard panic hook
<img width="1178" alt="image" src="https://github.com/ratatui-org/ratatui/assets/381361/7052fad1-ea27-4bad-8fc4-a25ec5c77786">

color-eyre panic hook
<img width="1196" alt="image" src="https://github.com/ratatui-org/ratatui/assets/381361/b94cec7b-41b1-4e06-90da-b59118880a36">

> One example for color_eyre, fine, all examples? Should be ratatui specific.

One common module used by all examples... this will be handled once, consistently the same way for all examples. There's no overhead to learning this or downside to it either.

```rust
mod common;

fn main() -> color_eyre::Result {
    common::install_hooks()?;
}
```

joshka replied:
> To that end, would having a `minimalist-example.rs` be useful?

Yup

joshka replied:
https://github.com/ratatui-org/ratatui/pull/1114

EdJoPaTo replied:
My point here is that a panic hook can be achieved with std tools without additional dependencies.

Having a panic hook is definitely a good thing for examples. I just don’t see the benefit of color_eyre. Especially since the panic message formatting was improved recently.

Personal opinion: either something is bundled with ratatui, like reexported, then it’s fine to be included in common examples. Or it should stay in a single example on how to use ratatui with color_eyre instead of the std lib way. Ratatui depends on std, but it does not on color_eyre.

I dislike examples that mash different things together as a default so users don’t learn the parts. Instead it’s a bundled up bunch hard to separate. In this case it hides the fact that a panic hook is working perfectly fine without color_eyre.
The common example should be a minimal example what’s needed for a general app. Additional things should add on to it. Having colored panic messages is an additional thing. Exactly like the bar chart is not part of the common file.

joshka replied:
> My point here is that a panic hook can be achieved with std tools without additional dependencies.

> Having a panic hook is definitely a good thing for examples. I just don’t see the benefit of color_eyre. Especially since the panic message formatting was improved recently.

The message format pales in comparison to the color eyre format. Color eyre has benefits listed above. Those benefits are useful to writing apps.

> Personal opinion: either something is bundled with ratatui, like reexported, then it’s fine to be included in common examples. Or it should stay in a single example on how to use ratatui with color_eyre instead of the std lib way. Ratatui depends on std, but it does not on color_eyre.

Opinion: examples should use things which are widely used in the rust economy system. Color eyre and anyhow which it is based on have a much more significant reach compared to Ratatui.

> I dislike examples that mash different things together as a default so users don’t learn the parts. Instead it’s a bundled up bunch hard to separate. In this case it hides the fact that a panic hook is working perfectly fine without color_eyre.

You have to write a panic hook and learn how to do that, this is just the best available version of that.

> The common example should be a minimal example what’s needed for a general app. Additional things should add on to it. Having colored panic messages is an additional thing. Exactly like the bar chart is not part of the common file.

To summarize your perspetive on this:
- you believe that only necessary things to demonstrate the library calls should be in examples
- you believe color eyre provides no benefit over the built in panic hooks and messages that are printed on crashes
- you believe that adding color eyre makes examples harder to understand

I disagree with all three statements. If I've straw-manned your arguments, that's not my intent.

My rationale for this is based on looking at a wide variety of apps built on ratatui, many of them "writing an app to learn rust", and seeing what people end up with. Example code is copied verbatim **regularly**. I've also interacted and provide support with ratatui users across:
- discord (4 different servers)
- Reddit
- Rust lang forums
- Many code repos looking at common code patterns, bugs, pain points, issues etc.
- Issues / discussion in the ratatui repo

Each of these have shown the need for some common patterns to help drive apps to be better.

I've also written a significant amount of docs for Ratatui and read the feedback from users about those docs.

People learn non linearly and hit examples directly for the thing they care about. Having different approaches per example leads to the good approaches being skipped when they're only covered in a single place. Doing the right thing across all the examples is the right thing to do here.

Thinking more about this, I think it's probably worth bringing the common module into each example rather than extracting it to a separate file, as this makes each example self contained even at the cost of duplication.

I believe strongly that using color eyre in all our examples is the right approach to use here for ratatui, even taking into consideration and understanding your opinions to the contrary.

On the opposite note, I believe you were right about not using the prelude in example code and that's something I'd like to roll out completely as part of revamping the examples. Not sure if you saw the message I sent you about the forum post I made on the rust users forum about this https://users.rust-lang.org/t/to-use-prelude-or-to-not-to-use-prelude-that-is-the-question/110855. The opinions were divided there, but there was enough evidence that using the prelude hampered the understanding of the code for readers to make that change.

joshka commented:
fix: tweaks

- Move constants closer to their usage
- add an exit field to app
- extract methods from the main loop to make the flow clearer
- move update handling to use a field on app instead of a variable in the loop
- rename last_tick to last_update to make it clearer what it controls

EdJoPaTo commented:
Given that there are other Repos with widget examples… how about a ratatui feature for an example base? Other widgets in other repos can also use them as a common base to demonstrate their widgets then.

joshka commented:
> Given that there are other Repos with widget examples… how about a ratatui feature for an example base? Other widgets in other repos can also use them as a common base to demonstrate their widgets then.

An example is just an app.
Adding code to make writing examples easy is adding code to make writing apps easy.
There shouldn't be an "example base", there should be functions that make using common parts of ratatui easy. Those changes are more difficult than cleaning up existing code to make it easier for new users to understand.

EdJoPaTo commented:
> > Given that there are other Repos with widget examples… how about a ratatui feature for an example base? Other widgets in other repos can also use them as a common base to demonstrate their widgets then.
> 
> An example is just an app. Adding code to make writing examples easy is adding code to make writing apps easy. There shouldn't be an "example base", there should be functions that make using common parts of ratatui easy. Those changes are more difficult than cleaning up existing code to make it easier for new users to understand.

This creates a common example base. My point is to make this `common.rs` reusable for other repos too, not just examples within the ratatui repo. That way they can also be more consistent.

joshka commented:
I think it might be better to incorporate that file in each example instead as it makes each example self contained as a single file.

EdJoPaTo commented on examples/barchart.rs:
```diff
@@ -14,289 +14,358 @@
     time::{Duration, Instant},
 };
 
+use common::Terminal;
```
Clarify where this comes from

```suggestion
use self::common::Terminal;
```

joshka commented:
![barchart](https://github.com/user-attachments/assets/daf7cdf0-fb2f-43b3-b460-a0a55c294b12)

kdheepak commented on examples/barchart.rs:
```diff
@@ -13,290 +13,320 @@
-            months: ["Mars", "Apr", "May", "Jun"],
+        Self {
+            should_exit: false,
+            data: Self::random_barchart_data(),
+            companies: Company::fake_companies(),
+            revenues: Revenues::fake_revenues(),
```
(nonblocking)

I have a minor preference for keeping all the data together. What do you think about `Self::fake_companies()` and `Self::fake_revenues()` instead and having the implementations for these functions live next to the implementation for `Self::random_barchart_data()`?

joshka replied:
I thnk you're right about moving these out of companies and revenues, but I land on them being bare functions instead. 765c50929

kdheepak reviewed (approved):
Looks much nicer to me!

joshka commented:
In the latest I simplified this a bunch more.
- Removed the Widget implementation from App and replaced it with just a plain render method that accepts Frame. This means that there's no need to split into two impl App blocks (which confused at least one newer rustacean)
- renamed common to terminal
- moved the color_eyre initialization into the terminal::init method
- removed all the animation (the animation demonstrates how to handle updates in ratatui apps, rather than just the widget)
- moved things around in layout / color etc.
- removed the legend (to avoid implying that the legend was part of the widget)

@EdJoPaTo aside from the color_eyre parts on this (which I'd suggest we agree to disagree on and move past), does this seem reasonable? I'd like to move forward with generally bringing each example up to scratch here using this pattrn. Once we work out #1180 and #1181, the terminal init and error handling parts can easily be replaced with whatever that comes up with, but those are a longer conversations that are currently stalled. Those concerns shouldn't block making the examples be consistent and useful for users.

joshka commented:
Small number tweak to avoid the 12.5M label looking weird

![barchart](https://github.com/user-attachments/assets/3111873b-578b-47f5-87c3-346b13172301)

joshka commented:
Split in two based on conversation with @kdheepak 
![barchart](https://github.com/user-attachments/assets/65fbc930-8222-42a6-8a7c-e62492fb4902)
![barchart-grouped](https://github.com/user-attachments/assets/8ae2f9ca-b368-411d-a1f8-8f05c0e3338a)

joshka commented:
Merging this as it's better, as it simplifies the main logic. Whether it's "best" is up for debate, but it's a demo of approximately what #1289 would look like (but with the methods in an example module)
